MessagePack cannot serialize BigInteger larger than 2^64-1

**Describe the bug**

<!-- A clear and concise description of what the bug is. -->
An [error](https://console.cloud.google.com/errors/detail/CJSquanBzMSL-AE;service=zeebe;filter=%5B%5D;time=P7D?project=camunda-cloud-240911) was reported: **RuntimeException: Failed to convert JSON to MessagePack.**

Caused by: `java.lang.IllegalArgumentException: MessagePack cannot serialize BigInteger larger than 2^64-1`

This is just another in a long list of issues with the MessagePack conversion:
- #7187
- #7025
- #5565
- #1441 

**To Reproduce**

<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample
- when using the Java client: https://github.com/zeebe-io/zeebe-test-template-java

-->
Just run the following: 

`zbctl create instance 123 --variables='{"mybigintistoolong":123456789012345678901234567890}'`
>Error: rpc error: code = Internal desc = Unexpected error occurred during the request processing: Failed to convert JSON to MessagePack

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->

User input should not lead to reported errors, but user command should be rejected with a helpful message.

**Log/Stacktrace**

<!-- If possible add the full stacktrace or Zeebe log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
 java.lang.RuntimeException: Failed to convert JSON to MessagePack
	at io.camunda.zeebe.protocol.impl.encoding.MsgPackConverter.convertToMsgPack(MsgPackConverter.java:80) ~[zeebe-protocol-impl-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.protocol.impl.encoding.MsgPackConverter.convertToMsgPack(MsgPackConverter.java:71) ~[zeebe-protocol-impl-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.RequestMapper.ensureJsonSet(RequestMapper.java:175) ~[zeebe-gateway-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.RequestMapper.toCreateProcessInstanceRequest(RequestMapper.java:113) ~[zeebe-gateway-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.EndpointManager.sendRequestWithRetryPartitions(EndpointManager.java:353) ~[zeebe-gateway-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.EndpointManager.createProcessInstance(EndpointManager.java:177) ~[zeebe-gateway-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.GatewayGrpcService.createProcessInstance(GatewayGrpcService.java:77) ~[zeebe-gateway-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.gateway.protocol.GatewayGrpc$MethodHandlers.invoke(GatewayGrpc.java:1597) ~[zeebe-gateway-protocol-impl-8.0.3.jar:8.0.3]
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:182) ~[grpc-stub-1.45.1.jar:1.45.1]
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35) ~[grpc-api-1.45.1.jar:1.45.1]
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23) ~[grpc-api-1.45.1.jar:1.45.1]
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:340) ~[grpc-core-1.45.1.jar:1.45.1]
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:866) ~[grpc-core-1.45.1.jar:1.45.1]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.45.1.jar:1.45.1]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133) ~[grpc-core-1.45.1.jar:1.45.1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
Caused by: java.lang.IllegalArgumentException: MessagePack cannot serialize BigInteger larger than 2^64-1
	at org.msgpack.core.MessagePacker.packBigInteger(MessagePacker.java:615) ~[msgpack-core-0.9.1.jar:0.9.1]
	at org.msgpack.jackson.dataformat.MessagePackGenerator.pack(MessagePackGenerator.java:230) ~[jackson-dataformat-msgpack-0.9.1.jar:0.9.1]
	at org.msgpack.jackson.dataformat.MessagePackGenerator.addValueToStackTop(MessagePackGenerator.java:591) ~[jackson-dataformat-msgpack-0.9.1.jar:0.9.1]
	at org.msgpack.jackson.dataformat.MessagePackGenerator.writeNumber(MessagePackGenerator.java:414) ~[jackson-dataformat-msgpack-0.9.1.jar:0.9.1]
	at com.fasterxml.jackson.core.JsonGenerator.copyCurrentEvent(JsonGenerator.java:2463) ~[jackson-core-2.13.2.jar:2.13.2]
	at com.fasterxml.jackson.core.JsonGenerator.copyCurrentStructure(JsonGenerator.java:2555) ~[jackson-core-2.13.2.jar:2.13.2]
	at io.camunda.zeebe.protocol.impl.encoding.MsgPackConverter.convert(MsgPackConverter.java:136) ~[zeebe-protocol-impl-8.0.3.jar:8.0.3]
	at io.camunda.zeebe.protocol.impl.encoding.MsgPackConverter.convertToMsgPack(MsgPackConverter.java:76) ~[zeebe-protocol-impl-8.0.3.jar:8.0.3]
	... 17 more 
```

</p>
</details>

**Environment:**
- Zeebe Version: 8.0.3, 8.0.4, etc.
- Both SaaS and SM

--- a/gateway/src/main/java/io/camunda/zeebe/gateway/RequestMapper.java
+++ b/gateway/src/main/java/io/camunda/zeebe/gateway/RequestMapper.java
@@ -189,8 +189,7 @@ public final class RequestMapper {
         return new UnsafeBuffer(MsgPackConverter.convertToMsgPack(value));
       } catch (final RuntimeException e) {
         final var cause = e.getCause();
-        if (cause instanceof JsonParseException) {
-          final var parseException = (JsonParseException) cause;
+        if (cause instanceof final JsonParseException parseException) {
 
           final var descriptiveException =
               new JsonParseException(
@@ -201,6 +200,9 @@ public final class RequestMapper {
 
           rethrowUnchecked(descriptiveException);
           return DocumentValue.EMPTY_DOCUMENT; // bogus return statement
+        } else if (cause instanceof IllegalArgumentException) {
+          rethrowUnchecked(cause);
+          return DocumentValue.EMPTY_DOCUMENT;
         } else {
           throw e;
         }

--- a/gateway/src/main/java/io/camunda/zeebe/gateway/grpc/GrpcErrorMapper.java
+++ b/gateway/src/main/java/io/camunda/zeebe/gateway/grpc/GrpcErrorMapper.java
@@ -76,6 +76,9 @@ public final class GrpcErrorMapper {
     } else if (error instanceof JsonParseException) {
       builder.setCode(Code.INVALID_ARGUMENT_VALUE).setMessage(error.getMessage());
       logger.debug("Expected to handle gRPC request, but JSON property was invalid", rootError);
+    } else if (error instanceof IllegalArgumentException) {
+      builder.setCode(Code.INVALID_ARGUMENT_VALUE).setMessage(error.getMessage());
+      logger.debug("Expected to handle gRPC request, but JSON property was invalid 123", rootError);
     } else if (error instanceof PartitionNotFoundException) {
       builder.setCode(Code.UNAVAILABLE_VALUE).setMessage(error.getMessage());
       logger.debug(

--- a/gateway/src/test/java/io/camunda/zeebe/gateway/RequestMapperTest.java
+++ b/gateway/src/test/java/io/camunda/zeebe/gateway/RequestMapperTest.java
@@ -18,6 +18,10 @@ public class RequestMapperTest {
   private static final String INVALID_VARIABLES =
       "{ \"test\": \"value\", \"error\": \"errorrvalue }";
 
+  // BigInteger larger than 2^64-1
+  private static final String BIG_INTEGER =
+      "{\"mybigintistoolong\": 123456789012345678901234567890}";
+
   @Test
   public void shouldThrowHelpfulExceptionIfJsonIsInvalid() {
     // when + then
@@ -26,5 +30,13 @@ public class RequestMapperTest {
         .hasMessageContaining("Invalid JSON", INVALID_VARIABLES)
         .cause()
         .isInstanceOf(JsonParseException.class);
+  }
+
+  @Test
+  public void shouldThrowHelpfulExceptionIfJsonHasBigInteger() {
+    // when + then
+    assertThatThrownBy(() -> RequestMapper.ensureJsonSet(BIG_INTEGER))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("MessagePack cannot serialize BigInteger larger than 2^64-1");
   }
 }

--- a/gateway/src/test/java/io/camunda/zeebe/gateway/grpc/GrpcErrorMapperTest.java
+++ b/gateway/src/test/java/io/camunda/zeebe/gateway/grpc/GrpcErrorMapperTest.java
@@ -144,4 +144,27 @@ final class GrpcErrorMapperTest {
       assertThat(event.getLevel()).isEqualTo(Level.DEBUG);
     }
   }
+
+  @Test
+  void shouldLogIllegalArgumentExceptionOnDebugWithBigIntTooLong() {
+    // given
+    try {
+      MsgPackConverter.convertToMsgPack("{\"mybigintistoolong\":123456789012345678901234567890}");
+      fail("Expected to throw exception");
+    } catch (final RuntimeException runtimeException) {
+      assertThat(runtimeException.getCause()).isInstanceOf(IllegalArgumentException.class);
+      final IllegalArgumentException exception =
+          (IllegalArgumentException) runtimeException.getCause();
+
+      // when
+      log.setLevel(Level.DEBUG);
+      final StatusRuntimeException statusException = errorMapper.mapError(exception, logger);
+
+      // then
+      assertThat(statusException.getStatus().getCode()).isEqualTo(Code.INVALID_ARGUMENT);
+      assertThat(recorder.getAppendedEvents()).hasSize(1);
+      final LogEvent event = recorder.getAppendedEvents().get(0);
+      assertThat(event.getLevel()).isEqualTo(Level.DEBUG);
+    }
+  }
 }

--- a/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/encoding/MsgPackConverter.java
+++ b/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/encoding/MsgPackConverter.java
@@ -77,7 +77,11 @@ public final class MsgPackConverter {
 
       return outputStream.toByteArray();
     } catch (final Exception e) {
-      throw new RuntimeException("Failed to convert JSON to MessagePack", e);
+      if (e instanceof IllegalArgumentException) {
+        throw new IllegalArgumentException("Failed to convert JSON to MessagePack", e);
+      } else {
+        throw new RuntimeException("Failed to convert JSON to MessagePack", e);
+      }
     }
   }
 

--- a/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CompleteJobTest.java
+++ b/qa/integration-tests/src/test/java/io/camunda/zeebe/it/client/command/CompleteJobTest.java
@@ -94,6 +94,21 @@ public final class CompleteJobTest {
   }
 
   @Test
+  public void shouldRejectIfVariableIsBigIntTooLong() {
+    // when
+    assertThatThrownBy(
+            () ->
+                CLIENT_RULE
+                    .getClient()
+                    .newCompleteCommand(jobKey)
+                    .variables("{\"mybigintistoolong\":123456789012345678901234567890}")
+                    .send()
+                    .join())
+        .isInstanceOf(ClientStatusException.class)
+        .hasMessageContaining("MessagePack cannot serialize BigInteger larger than 2^64-1");
+  }
+
+  @Test
   public void shouldRejectIfJobIsAlreadyCompleted() {
     // given
     CLIENT_RULE.getClient().newCompleteCommand(jobKey).send().join();

